### PR TITLE
ENH: `stats.qmc`: faster hypercube point comparison and scrambling doc

### PR DESCRIPTION
**Pull request by tupui**
Few minor things on QMC:

* Fix scrambling doc in Sobol'. It's not using Owen (NUS) but something from Matousek (LMS+shift). This came from a personal communication with Pierre L'Ecuyer.
* Faster (and clearer) point within hypercube check.
* Refactor usage of deprecated `assert_almost_equal` and `assert_array_almost_equal`.
* Uniformize a bit more the bibliography entries.

### CODE REVIEW

**Comment by tupui** [comment]
@mdhaber can you take a look? I moved some code over from the Lloyd PR here and took the opportunity to update very minor things.

File: scipy/stats/_qmc.py [hunk 1]
```diff
@@ -164,7 +164,7 @@ def scale(
 
     if not reverse:
         # Checking that sample is within the hypercube
-        if not (np.all(sample >= 0) and np.all(sample <= 1)):
+        if not ((sample.max() <= 1.) and (sample.min() >= 0.)):
```
Comment (mdhaber) [inline scipy/stats/_qmc.py:167]: Can't review without making suggestions, right?

```suggestion
        if (sample.max() > 1.) or (sample.min() < 0.):
```

Comment (tupui) [reply scipy/stats/_qmc.py:167]: Haha sure!

Indeed, better than having the not.

File: scipy/stats/_qmc.py [hunk 2]
```diff
@@ -294,7 +294,7 @@ def discrepancy(
     if not sample.ndim == 2:
         raise ValueError("Sample is not a 2D array")
 
-    if not (np.all(sample >= 0) and np.all(sample <= 1)):
+    if not ((sample.max() <= 1.) and (sample.min() >= 0.)):
```
Comment (tupui) [inline scipy/stats/_qmc.py:297]:
```suggestion
    if (sample.max() > 1.) or (sample.min() < 0.):
```

File: scipy/stats/_qmc.py [hunk 3]
```diff
@@ -357,7 +357,7 @@ def update_discrepancy(
     if not sample.ndim == 2:
         raise ValueError('Sample is not a 2D array')
 
-    if not (np.all(sample >= 0) and np.all(sample <= 1)):
+    if not ((sample.max() <= 1.) and (sample.min() >= 0.)):
```
Comment (tupui) [inline scipy/stats/_qmc.py:360]:
```suggestion
    if (sample.max() > 1.) or (sample.min() < 0.):
```

File: scipy/stats/_qmc.py [hunk 4]
```diff
@@ -794,7 +794,7 @@ class Halton(QMCEngine):
        points in evaluating multi-dimensional integrals", Numerische
        Mathematik, 1960.
     .. [2] A. B. Owen. "A randomized Halton algorithm in R",
-       arXiv:1706.02808, 2017.
+       :arxiv:`1706.02808`, 2017.
```
Comment (mdhaber) [inline scipy/stats/_qmc.py:797]: Before:
image

Now:
image

Good.

File: scipy/stats/_qmc.py (hunk 1 above)
Comment (mdhaber) [reply scipy/stats/_qmc.py:167]: Shall I wait for this to merge or go for it?

**Review by mdhaber — APPROVED** [review]
Looks like an improvement to me. I'll go ahead and merge now if you like or wait for you to tweak the within-hypercube checks.

**Comment by tupui** [comment]
> Looks like an improvement to me. I'll go ahead and merge now if you like or wait for you to tweak the within-hypercube checks.

Thanks, I just pushed your suggestion so yes you can go ahead.

**Comment by mdhaber** [comment]
OK I'll merge when tests are done. Or if I notice, when tests pass on Meson. I don't imagine this will be platform-dependent.

**Comment by mdhaber** [comment]
Thanks @tupui!

**Pull request merged**